File: src/main/kotlin/davidoh/profiling/SimpleProfiler.kt
/**
Description : A section-based runtime profiler.
Author      : david.ohana@ibm.com
License     : Apache v2
 **/

package davidoh.profiling


interface Profiler {
    fun startSection(name: String)
    fun endSection(name: String = "")
    fun report(periodSec: Int = 30): Boolean
}

/**
 * A no-op profiler. Use it instead of SimpleProfiler in case you want to disable profiling.
 */
class NullProfiler : Profiler {
    override fun startSection(name: String) {
    }

    override fun endSection(name: String) {
    }

    override fun report(periodSec: Int) = false
}

class SimpleProfiler(
    val resetAfterSampleCount: Long = 0,
    val enclosingSectionName: String = "total",
    val printer: (String) -> Unit = ::println,
) : Profiler {
    private val sectionToStats = mutableMapOf<String, ProfiledSectionStats>()
    private var lastReportTimeMillis = System.currentTimeMillis()
    private var lastStartedSectionName = ""
    private var nesting = 0

    /**
     * Start measuring a section.
     */
    override fun startSection(name: String) {
        require(name.isNotBlank())
        lastStartedSectionName = name

        val section = sectionToStats.getOrPut(name, { ProfiledSectionStats(name, nesting) })
        if (section.startTimeNano != 0L)
            throw IllegalArgumentException("Section $name already started")
        nesting++
        section.startTimeNano = System.nanoTime()
    }

    /**
     * End measuring a section.
     *
     * @param name name of the section to end. When empty or omitted, the last started section will be ended.
     */
    override fun endSection(name: String) {
        val nowNano = System.nanoTime()

        nesting--

        var sectionName = name
        if (name == "")
            sectionName = lastStartedSectionName

        check(sectionName.isNotBlank())

        val section = sectionToStats[sectionName]
            ?: throw IllegalArgumentException("Section $name does not exist")

        if (section.startTimeNano == 0L)
            throw IllegalStateException("section $name was not started")

        val tookNano = nowNano - section.startTimeNano
        if (resetAfterSampleCount > 0 && section.sampleCount == resetAfterSampleCount) {
            section.sampleCountBatch = 0
            section.totalTimeNanoBatch = 0
        }
        section.sampleCount++
        section.totalTimeNano += tookNano
        section.sampleCountBatch++
        section.totalTimeNanoBatch += tookNano
        section.startTimeNano = 0
    }

    /**
     * Print results using [printer] function. By default prints to stdout.
     */
    override fun report(periodSec: Int): Boolean {
        if (System.currentTimeMillis() - lastReportTimeMillis < periodSec * 1000)
            return false

        var enclosingTimeNano = 0L
        if (enclosingSectionName.isNotEmpty()) {
            val enclosingSection = sectionToStats[enclosingSectionName]
            if (enclosingSection != null)
                enclosingTimeNano = enclosingSection.totalTimeNano
        }

        val includeBatchRates = resetAfterSampleCount > 0
        val text = sectionToStats.values
            .sortedByDescending { it.totalTimeNano }
            .map { it.toString(enclosingTimeNano, includeBatchRates) }
            .joinToString(System.lineSeparator())
        printer(text)

        lastReportTimeMillis = System.currentTimeMillis()
        return true
    }
}


data class ProfiledSectionStats(
    val sectionName: String,
    var nesting: Int = 0,
    var startTimeNano: Long = 0,
    var sampleCount: Long = 0,
    var totalTimeNano: Long = 0,
    var sampleCountBatch: Long = 0,
    var totalTimeNanoBatch: Long = 0,
) {
    fun toString(enclosingTimeNano: Long, includeBatchRates: Boolean): String {
        var tookSecText = (totalTimeNano / 1000000000.0).format(6, 2) + " s"
        if (enclosingTimeNano > 0) {
            val contributionPercent = (100.0 * totalTimeNano / enclosingTimeNano).format(3, 2)
            tookSecText += " ($contributionPercent%)"
        }
        val sampleCountText = sampleCount.format(12)
        var msPerKSamples = (totalTimeNano / (1000.0 * sampleCount.toDouble())).format(6, 2)
        var samplesPerSec = (sampleCount.toDouble() / totalTimeNano * 1000000000).format(10, 2)
        if (includeBatchRates) {
            msPerKSamples += " (${(totalTimeNanoBatch / (1000.0 * sampleCountBatch.toDouble())).format(6, 2)})"
            samplesPerSec += " (${(sampleCountBatch.toDouble() / totalTimeNanoBatch * 1000000000).format(10, 2)})"
        }

        val nestingText = " ".repeat(nesting)
        val name = sectionName.padEnd(20 - nesting)
        return "$nestingText$name: took $tookSecText, $sampleCountText samples, " +
                "$msPerKSamples ms / 1000 samples, $samplesPerSec hz"
    }
}

fun Number.format(
    charsAboveZero: Int = 1, charsBelowZero: Int = 0,
    isSigned: Boolean = false, leadingZeros: Boolean = false
): String {
    val isInteger = this is Long || this is Int || this is Short

    if (isInteger && charsBelowZero > 0)
        return this.toDouble().format(charsAboveZero, charsBelowZero, isSigned, leadingZeros)

    val leadingZero = if (leadingZeros) "0" else ""
    val sign = if (isSigned) "+" else ""
    val length = charsAboveZero + charsBelowZero + (if (charsBelowZero > 0) 1 else 0) + sign.length
    val fraction = if (isInteger) "" else ".$charsBelowZero"
    val type = if (isInteger) "d" else "f"
    val template = "%$sign$leadingZero,$length$fraction$type"
    return template.format(this)
}

File: python/simple_profiler.py
"""
Description : A simple section-based performance profiler
Author      : David Ohana
Author_email: david.ohana@ibm.com
License     : MIT
"""
import os
import time

from abc import ABC, abstractmethod


class Profiler(ABC):

    @abstractmethod
    def start_section(self, section_name: str):
        pass

    @abstractmethod
    def end_section(self, section_name=""):
        pass

    @abstractmethod
    def periodic_report(self):
        pass

    @abstractmethod
    def print_results(self):
        pass


class NullProfiler(Profiler):
    def start_section(self, section_name: str):
        pass

    def end_section(self, section_name=""):
        pass

    def periodic_report(self):
        pass

    def print_results(self):
        pass


class SimpleProfiler(Profiler):
    def __init__(self, reset_after_sample_count=0, enclosing_section_name="", printer=print, report_sec=30):
        self.printer = printer
        self.enclosing_section_name = enclosing_section_name
        self.reset_after_sample_count = reset_after_sample_count
        self.report_sec = report_sec

        self.section_to_stats = {}
        self.last_report_timestamp_sec = time.time()
        self.last_started_section_name = ""

    def start_section(self, section_name: str):
        if not section_name:
            raise ValueError("Section name is empty")
        self.last_started_section_name = section_name

        section = self.section_to_stats.get(section_name, None)
        if section is None:
            section = ProfiledSectionStats(section_name)
            self.section_to_stats[section_name] = section

        if section.start_time_sec != 0:
            raise ValueError(f"Section {section_name} is already started")

        section.start_time_sec = time.time()

    def end_section(self, name=""):
        now = time.time()

        section_name = name
        if not name:
            section_name = self.last_started_section_name

        if not section_name:
            raise ValueError("Neither section name is specified nor a section is started")

        section: ProfiledSectionStats = self.section_to_stats.get(section_name, None)
        if section is None:
            raise ValueError(f"Section {section_name} does not exist")

        if section.start_time_sec == 0:
            raise ValueError(f"Section {section_name} was not started")

        took_sec = now - section.start_time_sec
        if self.reset_after_sample_count > 0 and section.sample_count == self.reset_after_sample_count:
            section.sample_count_batch = 0
            section.total_time_sec_batch = 0

        section.sample_count += 1
        section.total_time_sec += took_sec
        section.sample_count_batch += 1
        section.total_time_sec_batch += took_sec
        section.start_time_sec = 0

    def print_results(self):
        enclosing_time_sec = 0
        if self.enclosing_section_name:
            enclosing_section: ProfiledSectionStats = self.section_to_stats.get(self.enclosing_section_name, None)
            if enclosing_section:
                enclosing_time_sec = enclosing_section.total_time_sec

        include_batch_rates = self.reset_after_sample_count > 0
        text = ""

        sections = self.section_to_stats.values()
        sorted_sections = sorted(sections, key=lambda it: it.total_time_sec, reverse=True)
        for section in sorted_sections:
            text += section.to_string(enclosing_time_sec, include_batch_rates) + os.linesep
        self.printer(text)

    def periodic_report(self):
        if time.time() - self.last_report_timestamp_sec < self.report_sec:
            return
        self.print_results()
        self.last_report_timestamp_sec = time.time()


class ProfiledSectionStats:
    def __init__(self, section_name, start_time_sec=0, sample_count=0, total_time_sec=0,
                 sample_count_batch=0, total_time_sec_batch=0):
        self.section_name = section_name
        self.start_time_sec = start_time_sec
        self.sample_count = sample_count
        self.total_time_sec = total_time_sec
        self.sample_count_batch = sample_count_batch
        self.total_time_sec_batch = total_time_sec_batch

    def to_string(self, enclosing_time_sec: int, include_batch_rates: bool):
        took_sec_text = f"{self.total_time_sec:>8.2f} s"
        if enclosing_time_sec > 0:
            took_sec_text += f" ({100 * self.total_time_sec / enclosing_time_sec:>6.2f})%"

        ms_per_k_samples = f"{1000000 * self.total_time_sec / self.sample_count: 7.2f}"
        samples_per_sec = f"{self.sample_count / self.total_time_sec: 15,.2f}"

        if include_batch_rates:
            ms_per_k_samples += f" ({1000000 * self.total_time_sec_batch / self.sample_count_batch: 7.2f})"
            samples_per_sec += f" ({self.sample_count_batch / self.total_time_sec_batch: 15,.2f})"

        return f"{self.section_name: <15}: took {took_sec_text}, " \
               f"{self.sample_count: >10,} samples, " \
               f"{ms_per_k_samples} ms per 1K samples, " \
               f"{samples_per_sec} hz"